examples/pytorch/sagpool/main.py

```diff
@@ -0,0 +1,191 @@
+                        choices=["hierarchical", "global"],
+                        help="model architecture")
+    parser.add_argument("--dataset_path", type=str, default="./dataset",
+                        help="path to dataset")
+    parser.add_argument("--conv_layers", type=int, default=3,
+                        help="number of conv layers")
+    parser.add_argument("--print_every", type=int, default=10,
+                        help="print trainlog every k epochs, -1 for silent training")
+    parser.add_argument("--num_trials", type=int, default=1,
+                        help="number of trials")
+    parser.add_argument("--output_path", type=str, default="./output")
+    
+    args = parser.parse_args()
+
+    # device
+    args.device = "cpu" if args.device == -1 else "cuda:{}".format(args.device)
+    if not torch.cuda.is_available():
+        logging.warning("CUDA is not available, use CPU for training.")
+        args.device = "cpu"
+
+    # print every
+    if args.print_every == -1:
+        args.print_every = args.epochs + 1
+
+    # paths
+    if not os.path.exists(args.dataset_path):
+        os.makedirs(args.dataset_path)
+    if not os.path.exists(args.output_path):
+        os.makedirs(args.output_path)
+    name = "Data={}_Hidden={}_Arch={}_Pool={}_WeightDecay={}_Lr={}.log".format(
+        args.dataset, args.hid_dim, args.architecture, args.pool_ratio, args.weight_decay, args.lr)
+    args.output_path = os.path.join(args.output_path, name)
+
+    return args
+
+
+def train(model:torch.nn.Module, optimizer, trainloader, device):
+    model.train()
+    total_loss = 0.
+    for batch in trainloader:
+        optimizer.zero_grad()
+        batch_graphs, batch_labels = batch
+        batch_graphs = batch_graphs.to(device)
+        batch_labels = batch_labels.to(device)
+        out = model(batch_graphs)
+        loss = F.nll_loss(out, batch_labels)
+        loss.backward()
+        optimizer.step()
+
+        total_loss += loss.item()
+    
+    return total_loss / len(trainloader.dataset)
+
+
+@torch.no_grad()
+def test(model:torch.nn.Module, loader, device):
+    model.eval()
+    correct = 0.
+    loss = 0.
+    num_graphs = len(loader.dataset)
+    for batch in loader:
+        batch_graphs, batch_labels = batch
+        batch_graphs = batch_graphs.to(device)
+        batch_labels = batch_labels.to(device)
+        out = model(batch_graphs)
+        pred = out.argmax(dim=1)
+        loss += F.nll_loss(out, batch_labels, reduction="sum").item()
+        correct += pred.eq(batch_labels).sum().item()
+    return correct / num_graphs, loss / num_graphs
+
+
+def main(args):
+    # Step 1: Prepare graph data and retrieve train/validation/test index ============================= #
+    dataset = LegacyTUDataset(args.dataset, raw_dir=args.dataset_path)
+
+    # add self loop. We add self loop for each graph here since the function "add_self_loop" does not
```

Comment:
@yzh119 Do you think we can support batched graph transformation in a more principled way?